hack/prom-metric-linter/metrics_collector.go

```diff
@@ -0,0 +1,102 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2023 Red Hat, Inc.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+
+	dto "github.com/prometheus/client_model/go"
+)
+
+// Metric represents a Prometheus metric
+type Metric struct {
+	Name string
+	Help string
+	Type string
+}
+
+// Extract the name, help, and type from the metrics doc file
+func ExtractMetrics(metricsContent string) ([]Metric, error) {
+	var metrics []Metric
+	re := regexp.MustCompile(`### (.*)\n(.*Type: (Counter|Gauge|Histogram|Summary)\.\n)?`)
+	matches := re.FindAllStringSubmatch(metricsContent, -1)
+	for _, match := range matches {
+		name := match[1]
+		help := ""
+		if len(match) > 2 {
+			help = match[2]
+		}
+		help = strings.TrimSpace(help)
+		typ := ""
+		if len(match) > 3 {
+			typ = match[3]
+		}
+		metrics = append(metrics, Metric{Name: name, Help: help, Type: typ})
+	}
+	if len(metrics) == 0 {
+		return nil, fmt.Errorf("no metrics found")
+	}
+	return metrics, nil
+}
+
+// Set the correct metric type for creating MetricFamily
+func CreateMetricFamily(m Metric) *dto.MetricFamily {
+	metricType := dto.MetricType_UNTYPED
+
+	switch m.Type {
+	case "Counter":
+		metricType = dto.MetricType_COUNTER
+	case "Gauge":
+		metricType = dto.MetricType_GAUGE
+	case "Histogram":
+		metricType = dto.MetricType_HISTOGRAM
+	case "Summary":
+		metricType = dto.MetricType_SUMMARY
+	}
+
+	return &dto.MetricFamily{
+		Name: &m.Name,
+		Help: &m.Help,
+		Type: &metricType,
+	}
+}
+
+// Read the metrics and parse them to a MetricFamily
+func ReadMetrics() ([]*dto.MetricFamily, error) {
+	metricsContent, err := os.ReadFile("/root/go/src/kubevirt.io/kubevirt/docs/metrics.md")
```

Comment:
TODO:  update path to be more relative path and contain only /docs/metrics.md if possible.